### PR TITLE
Documentation improvements

### PR DESCRIPTION
**Pull request by nariman87**
## Context for changes

Major [readthedocs](shttps://flamingpy.readthedocs.io/en/code-quality/index.html) improvements as listed below and some minor `README.md` updates. 

- **API details and inheritance diagrams should be now correctly displayed**: 
    However, inheritance diagrams are limited to what was instructed for classes in original Python files. See for example [fp. codes](https://flamingpy.readthedocs.io/en/code-quality/source/fp.codes.html) the only package where two top classes were specified. See [here](https://quantum-jet.readthedocs.io/en/stable/code/jet_gate.html) and [here](https://www.sphinx-doc.org/en/master/usage/extensions/inheritance.html) for example on how to set up your classes for the best inheritance diagram experience.
- "Edit on Github" links were fixed
- The general style and section structures made more consistent with the company requirements and other packages such s StrawberryFields.
- Fixed the documentation links in `README.md`  
- All C++ imports are now placed within a try block to avoid interrupting non-compiled installations, such as the one currently used by readthedocs.
- Updated `doc/conf.py`, `doc/dev_requirements.txt`, and `doc/Makefile`, among other minor file changes, to achieve the above. 
- **Relative path cannot be used in `README.md` logos**:
    Was replaced with Github-hosted links.
- Code coverage and quality scores were improved to arguably acceptable levels.
  

## Example usage and tests

- [x] Not Applicable

## Performance results justifying changes

- [x] Not Applicable

## Workflow actions and tests

- [x] Not Applicable

## Expected benefits and drawbacks

**Expected benefits:**
- More consistent and modern-looking documentation when compared to other Xanadu and open-source products.
- The codefactor quality score is now increased to `A-`.

**Possible drawbacks:**
- For now, we are ignoring C++ APIs.

## Related Github issues

- [x] Not Applicable

## Checklist and integration statements

- [x] My Python and C++ codes follow the coding and commenting styles of this project as indicated by existing files. Specifically, the changes conform to given `black`, `docformatter` and `pylint` configurations. 
- [x] I have performed a self-review of these changes. I have checked my code and corrected misspellings to the best of my capacity. I also confirm that I have already merged other branches into this branch as required.
- [x] I have added context for corresponding changes in documentation and README.md as needed.
- [x] I have added new workflow CI tests for corresponding changes and these pass locally for me.
- [ ] I have updated `_version.py` based on [semantic versioning](https://semver.org/). I recognize that the developers may create a Special Release including my changes. **I suggest a Special Release PR is essential immediately following these changes**.

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -11,9 +11,17 @@
...
 import flamingpy
+
+# defaults
+TOL = 1e-5
+
+
+@pytest.fixture(scope="session")
+def tol():
+    """Numerical tolerance for equality tests."""
+    return float(os.environ.get("TOL", TOL))
```
Comment (ilan-tz) [inline tests/conftest.py:27]: Is this future-proofing (I don't think it's currently used anywhere)?

**Review by ilan-tz — CHANGES_REQUESTED** [review]
Great improvements to the docs -- aesthetic improvements, will align better with other Xanadu software, and readies docs for future changes. Just a few small questions.

File: README.md [hunk 2]
```diff
@@ -144,16 +150,17 @@ If you are having issues, please let us know by posting the issue on our GitHub
 
 You can also start a general discussion and connect with our community members in our [Discussions Page](https://github.com/XanaduAI/flamingpy/discussions). 
 
+
 ## Attribution for authors
 
-FlamingPy is the work of [many contributors](https://github.com/XanaduAI/ft-stack/graphs/contributors). 
+FlamingPy is the work of [many contributors](https://github.com/XanaduAI/ft-stack/graphs/contributors). If you are doing research using FlamingPy, please cite our paper below:
```
Comment (ilan-tz) [inline README.md:156]:
```suggestion
FlamingPy is the work of [many contributors](https://github.com/XanaduAI/flamingpy/graphs/contributors). If you are doing research using FlamingPy, please cite our paper below:

```

File: README.md [hunk 3]
```diff
@@ -144,16 +150,17 @@ If you are having issues, please let us know by posting the issue on our GitHub
...
-FlamingPy is the work of [many contributors](https://github.com/XanaduAI/ft-stack/graphs/contributors). 
+FlamingPy is the work of [many contributors](https://github.com/XanaduAI/ft-stack/graphs/contributors). If you are doing research using FlamingPy, please cite our paper below:
 
-If you are doing research using FlamingPy, please cite our paper below:
 
 > Ilan Tzitrin, Takaya Matsuura, Rafael N. Alexander, Guillaume Dauphinais, J. Eli Bourassa, Krishna K. Sabapathy, Nicolas C. Menicucci, and Ish Dhand,
 > Fault-Tolerant Quantum Computation with Static Linear Optics, PRX Quantum, Vol. 2, No. 4, 2021, 
 > [DOI:10.1103/prxquantum.2.040353](http://dx.doi.org/10.1103/PRXQuantum.2.040353) 
 
+In addition to the authors above, the developers would like to thank Michael Vasmer and Ashlesha Patil for their contributions to the pre-release project.
```
Comment (ilan-tz) [inline README.md:163]:
```suggestion
In addition to the authors above, the developers would like to thank Sanchit Bapat, Ashlesha Patil, Michael Vasmer, and Trevor Vincent for their contributions to the pre-release project. 

```

File: doc/index.rst [hunk 4]
```diff
@@ -22,24 +99,42 @@ Features
...
+
+FlamingPy is the work of `many contributors <https://github.com/XanaduAI/ft-stack/graphs/contributors>`__. The developers would like to specifically thank Michael Vasmer and Ashlesha Patil for their contributions to the pre-release project. 
+
+If you are doing research using FlamingPy, please cite our paper below:
+
+    Ilan Tzitrin, Takaya Matsuura, Rafael N. Alexander, Guillaume Dauphinais, J. Eli Bourassa, Krishna K. Sabapathy, Nicolas C. Menicucci, and Ish Dhand,
+    Fault-Tolerant Quantum Computation with Static Linear Optics, PRX Quantum, Vol. 2, No. 4, 2021, 
+    `DOI:10.1103/prxquantum.2.040353 <http://dx.doi.org/10.1103/PRXQuantum.2.040353>`__ 
+
+In addition to the authors above, the developers would like to thank Michael Vasmer and Ashlesha Patil for their contributions to the pre-release project.
```
Comment (ilan-tz) [inline doc/index.rst:113]:
```suggestion
In addition to the authors above, the developers would like to thank Sanchit Bapat, Ashlesha Patil, Michael Vasmer, and Trevor Vincent for their contributions to the pre-release project.
```

File: doc/development/guide_for_devs.rst [hunk 5]
```diff
@@ -10,13 +10,21 @@ FlamingPy requires the following Python version to be installed:
...
 * `pandas <https://pandas.pydata.org/>`_ >= 1.2.1
+* `retworkx <https://qiskit.org/documentation/retworkx/>`_ >= 0.10.2
 * `scipy <https://scipy.org/>`_ >= 1.6.0
-
+* `sphinx <https://www.sphinx-doc.org/en/master/index.html>`_ >= 4.3.1
+* `sphinxcontrib-bibtex <https://sphinxcontrib-bibtex.readthedocs.io/en/latest/>`_ >= 0.4.2
+* `sphinx-autodoc-typehints <https://pypi.org/project/sphinx-autodoc-typehints/>`_ >= 1.10.3
+* `sphinx-automodapi <https://sphinx-automodapi.readthedocs.io/en/latest/>`_ >= 0.13
+* `sphinx-copybutton <https://sphinx-copybutton.readthedocs.io/en/latest/>`_ >= 0.4
+* `toml <https://pypi.org/project/toml/>`_ >= 0.10.2
```
[the remark this answers is not in the document]
Comment (nariman87) [reply doc/development/guide_for_devs.rst:27]: thanks, I'll update `Guide for developers` and `Build Documentation` soon.

File: flamingpy/__init__.py [hunk 6]
```diff
@@ -25,8 +26,12 @@
 import matplotlib
 import pandas
 
-import flamingpy.cpp.lemonpy as lp
-import flamingpy.cpp.cpp_mc_loop as cmc
+try:
+    import flamingpy.cpp.lemonpy as lp
+    import flamingpy.cpp.cpp_mc_loop as cmc
+except ImportError:
+    warnings.warn("failed to import flamingpy.cpp libraries", ImportWarning)
```
Comment (ilan-tz) [inline flamingpy/__init__.py:33]:
```suggestion
    warnings.warn("Failed to import flamingpy.cpp libraries.", ImportWarning)
```

File: flamingpy/benchmarks/simulations.py [hunk 7]
```diff
@@ -22,7 +23,11 @@
 from flamingpy.decoders.decoder import correct
 from flamingpy.cv.ops import CVLayer
 from flamingpy.cv.macro_reduce import BS_network, reduce_macro_and_simulate
-import flamingpy.cpp.cpp_mc_loop as cmc
+
+try:
+    import flamingpy.cpp.cpp_mc_loop as cmc
+except ImportError:
+    warnings.warn("failed to import flamingpy.cpp.cpp_mc_loop library", ImportWarning)
```
Comment (ilan-tz) [inline flamingpy/benchmarks/simulations.py:30]:
```suggestion
    warnings.warn("Failed to import flamingpy.cpp.cpp_mc_loop library.", ImportWarning)
```

File: flamingpy/decoders/mwpm/lemon.py [hunk 8]
```diff
@@ -12,9 +12,14 @@
...
 """Define max_weight_matching based on the lemonpy module."""
+import warnings
+
 import networkx as nx
 
-import flamingpy.cpp.lemonpy as lp
+try:
+    import flamingpy.cpp.lemonpy as lp
+except ImportError:
+    warnings.warn("failed to import flamingpy.cpp.lemonpy library", ImportWarning)
```
Comment (ilan-tz) [inline flamingpy/decoders/mwpm/lemon.py:22]:
```suggestion
    warnings.warn("Failed to import flamingpy.cpp.lemonpy library.", ImportWarning)
```

File: flamingpy/__init__.py [hunk 9]
```diff
@@ -25,8 +26,12 @@
...
+
```
[the remark this answers is not in the document]
Comment (nariman87) [reply flamingpy/__init__.py:34]: As discussed, these warnings make sense for some common installation scenarios, so let's keep those.

**Comment by ilan-tz** [comment]
One thing I just noticed - some of the links in the Python API do not work (they are supposed to link to various docstrings)

File: tests/conftest.py (hunk 1 above)
Comment (nariman87) [reply tests/conftest.py:27]: Yes, exactly, future-proofing.

File: doc/conf.py [hunk 10]
```diff
@@ -44,9 +42,10 @@ def __getattr__(cls, name):
 author = "Xanadu Inc."
 
 # The full version, including alpha/beta/rc tags.
-# with open("../flamingpy/_version.py") as f:
-#    release = f.readlines()[-1].split()[-1].strip("\"'")
-release = fp.__version__
+with open("../flamingpy/_version.py") as f:
+    release = f.readlines()[-1].split()[-1].strip("\"'")
+# import flamingpy as fp
+# release = fp.__version__
```
[the remark this answers is not in the document]
Comment (nariman87) [reply doc/conf.py:48]: Yup, I forgot to explain this during repo discussions the other day. 

The current method has the advantage of not needing to import flamingpy, which allows us to remove a couple of dependencies from doc/dev_requirements.txt. Although, we may have to import flamingly in the future for other purposes either way.

File: doc/source/fp.rst [hunk 11]
```diff
@@ -1,10 +1,27 @@
 fp
 ==
 
-This and the following sections contain the Python API details for FlamingPy.
+In the following sections, we will detail the Python APIs for FlamingPy package.
```
Comment (ilan-tz) [inline doc/source/fp.rst:4]:
```suggestion
In the following sections, we will detail the Python APIs for FlamingPy. 
```

**Review by ilan-tz — APPROVED** [review]

**Pull request merged**